Error messages

gasche opened the pull request:
The way the error messages are obtained is by asking the parser generator for the list of its error states (and the inputs that lead it there, in a sense), and writing a message for each different case by hand. There are 32 error states in the current grammar, so writing 32 messages was a tad tedious, but then I got lost on the green line tonight so I had plenty of time on my hands.

The plumbing parts where I obtain the automaton error state number is admittedly rather arcane, as well as the build-system plumbing to obtain the error state information, compile the hand-written messages to an OCaml function, and update the messages file when the grammar change. But I think the result is relatively neat:

```
File "test.sou", line 3, characters 8-13, parsing error:
>   const limit 10
Parsing an instruction, we parsed "const <var>" so far; the equal sign
"=" is now expected to construct a constant declaration
"const <var> = <expr>".
```

```
File "test.sou", line 6, characters 0-9, parsing error:
> loop_body
Parsing an instruction, we parsed an identifier so far (variable or label). 
- if this is a label declaration, we expect a semicolon: "<label>:"
- if this is an assignment, we expect a left arrow: "<var> <- <expression>"
```

```
File "test.sou", line 9, characters 10-11, parsing error:
>   i <- (i + (1 + 2))
Parsing an expression, we parsed "( <arg> <op>" so far; an argument
(variable or literal value) is now expected to construct an expression
"( <arg> <op> <arg> )".
```

```
File "test.sou", line 9, characters 12-13, parsing error:
>   i <- (i + 1 + 2)
Parsing an expression, we parsed "( <arg> <op> <arg>" so far;
a closing parenthesis ")" is now expected.
```

gasche committed:
use Menhir's support for error messages

The way the error messages are generated is by asking Menhir to
produce a description of the possible error modes of the grammar in
a file parser.messages (from parser.mly) with placeholder for error
messages. Then I fill the placeholder with Gabriel-generater error
messages, based on Menhir's explanations of the error situation
(in comments). Then I can ask Menhir to generate the file
parser_messages.ml who maps an error state (of the parsing automaton)
to the right message.

Finally, a rather tricky bit of Menhir machinery needs to be done to
obtain the error state in case of a parsing error. This goes in
parse.ml, and it is done by carefully reading the Menhir manual.

gasche committed:
quote the wrong line

together with the previous commits, fixes issue #15

gasche committed:
add a .messages.update rule

Now "ocamlbuild parser.messages.update" will update the source
parser.messages file after parser changes. To be used cautiously.

o- commented on parse.ml:
```diff
@@ -9,9 +9,44 @@ let split_annotations program =
 let parse_string str =
   let lexbuf = Lexing.from_string str in
-  split_annotations (Parser.program Lexer.token lexbuf)
+  parse lexbuf
```
I know it's actually my leftovers -- but if you don't mind it would be nice to somehow have better errors for parse_string too

gasche replied:
Indeed, I can work on that. `parse_string` already uses the same auxiliary function. The reason why I haven't shared the error-reporting between `parse_file` and `parse_string` is that I first tried that, and that the `exit <error-code>` things break the testsuite (they stop the tests instead of continuing with other tests) while raising an uncaught exception does not. So we need something a bit more subtle, where the error-message-printing is done in both cases, but only one of them aborts. (Or maybe none, and then the users of `parse_file` with also die with an uncaught exception, which is maybe just fine.)

o- reviewed (commented):
Don't exactly understand the mechanism. But I guess that's ok. I love the result.

gasche committed:
clarify the Parse interface

The handling of parsing errors by `Parse.parse_*` functions was
inconsistent: `parse_file` would display a user-intended parsing error
on stderr and exit, while `parse_string` would just raise an
exception.

Now all parsing function raise exceptions on error, and the `Parse`
module also provides a `report_error` function for other modules to
explicitly request the user-friendly error on stderr. The possible
exceptions that can be raised have been consolidated in a single
type -- any other exception raised by parsing functions is a bug that
should be fixed.

gasche commented:
@o- I tried to clarify the `Parse` interface in cd1364258878e1903e7fb21907d88b4bc4bfecec, tell me what you think.

P.S.: `.mli` files are interface files that describe the functions exposed by a given module. Besides documentation, they also serve as access control: only the types and values mentioned in the interface are accessible from outside. Large OCaml projects prefer to have `.mli` files for all `.ml` files, but it makes sense to grow them on-demand (when we start feeling that an interface boundary should be better specified) for a young project.

gasche commented:
I decided to go ahead and merge given that I have another parser change in the pipeline -- support for empty lines.

o- commented:
oh wow, that's now the de-luxe version :) thanks
